src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastCache.cs

```diff
@@ -0,0 +1,461 @@
+        // This is all we need to implement TryGet.
+        private static int[]? s_table;
+#else
+
+  #if DEBUG
+        private const int INITIAL_CACHE_SIZE = 8;    // MUST BE A POWER OF TWO
+        private const int MAXIMUM_CACHE_SIZE = 512;  // make this lower than release to make it easier to reach this in tests.
+  #else
+        private const int INITIAL_CACHE_SIZE = 128;  // MUST BE A POWER OF TWO
+        private const int MAXIMUM_CACHE_SIZE = 4096; // 4096 * sizeof(CastCacheEntry) is 98304 bytes on 64bit. We will rarely need this much though.
+  #endif // DEBUG
+
+        private const int VERSION_NUM_SIZE = 29;
+        private const uint VERSION_NUM_MASK = (1 << VERSION_NUM_SIZE) - 1;
+
+        // A trivial 2-elements table used for "flushing" the cache. Nothing is ever stored in this table.
+        // It is required that we are able to allocate this.
+        private static int[] s_sentinelTable = CreateCastCache(2, throwOnFail: true)!;
+
+        // when flushing, remember the last size.
+        private static int s_lastFlushSize = INITIAL_CACHE_SIZE;
+
+        // The actual storage.
+        // Initialize to the sentinel in DEBUG as if just flushed, to ensure the sentinel can be handled in Set.
+        private static int[] s_table =
+  #if !DEBUG
+            CreateCastCache(INITIAL_CACHE_SIZE) ??
+  #endif
+            s_sentinelTable;
+
+#endif // CORECLR
+
+        private const int BUCKET_SIZE = 8;
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct CastCacheEntry
+        {
+            // version has the following structure:
+            // [ distance:3bit |  versionNum:29bit ]
+            //
+            // distance is how many iterations the entry is from it ideal position.
+            // we use that for preemption.
+            //
+            // versionNum is a monotonically increasing numerical tag.
+            // Writer "claims" entry by atomically incrementing the tag. Thus odd number indicates an entry in progress.
+            // Upon completion of adding an entry the tag is incremented again making it even. Even number indicates a complete entry.
+            //
+            // Readers will read the version twice before and after retrieving the entry.
+            // To have a usable entry both reads must yield the same even version.
+            //
+            internal uint _version;
+            internal nuint _source;
+            // pointers have unused lower bits due to alignment, we use one for the result
+            internal nuint _targetAndResult;
+
+            internal void SetEntry(nuint source, nuint target, bool result)
+            {
+                _source = source;
+                _targetAndResult = target | (nuint)(result ? 1 : 0);
+            }
+        }
+
+#if TARGET_64BIT
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong RotateLeft(ulong value, int offset)
```

Comment:
This should be deleted and replaced by `BitOperations.RotateLeft`.

Reply:
I've introduced this because NativeAOT was not building in the test configuration that was missing a lot of things - like Volatile, Interlocked, BitOperations. 
Later I moved custom implementations under Test.CoreLib, but missed this one.

Reply:
We may want to omit this cache for Test.CoreLib. Test.CoreLib has simplistic implementation of number of other subsystems.

Reply:
At this point it would be easy to move these helpers to Test.CoreLib - as custom BitOperations implementations. Then we can have the cache in Test>CorLib and have extra test coverage.

However, if having a minimal implementation is the point, then I can instead add a Test.CoreLib - specific implementation of the cache that does not do anything.

Reply:
> Then we can have the cache in Test.CorLib and have extra test coverage.

There are no relevant tests running against Test.CorLib. All tests that matter for this run against real CoreLib.

Reply:
I reverted additional API implementations for Test.CorLib and added a trivial no-op cache instead.